app/lib/activitypub/activity/create.rb

```diff
@@ -245,6 +252,15 @@ def skip_download?
     @skip_download ||= DomainBlock.find_by(domain: @account.domain)&.reject_media?
   end
 
+  def invalid_origin?(url)
+    return true if unsupported_uri_scheme?(url)
+
+    needle   = Addressable::URI.parse(url)
+    haystack = Addressable::URI.parse(@account.uri)
```

Comment:
this may be more clear if you do the \.host calls here, so it's:

```ruby
object_domain = Addressable::URI.parse(url).host
actor_domain  = Addressable::URI.parse(@account.uri).host
```